rig/place_and_route/place/utils.py

```diff
@@ -86,3 +91,151 @@ def apply_reserve_resource_constraint(machine, constraint):
+
+            # ...deal with other types of constraint...
+
+            # ...perform placement...
+
+            finalise_same_chip_constraints(substitutions, placements)
+            return placements
+
+    Note that this function does not modify its arguments but rather returns
+    new copies of the structures supplied.
+
+    Parameters
+    ----------
+    vertices_resources : {vertex: {resource: quantity, ...}, ...}
+    nets : [:py:class:`~rig.netlist.Net`, ...]
+    constraints : [constraint, ...]
+
+    Returns
+    -------
+    (vertices_resources, nets, constraints, substitutions)
+        The vertices_resources, nets and constraints values contain modified
+        copies of the supplied data structures modified to contain a single
+        vertex in place of the individual constrained vertices.
+
+        substitutions is a list of :py:class:`MergedVertex` objects which
+        resulted from the combining of the constrained vertices. The order of
+        the list is the order the substitutions were carried out. The
+        :py:func:`finalise_same_chip_constraints` function can be used to
+        expand a set of substitutions.
+    """
+    # Make a copy of the basic structures to be modified by this function
+    vertices_resources = vertices_resources.copy()
+    nets = nets[:]
+    constraints = constraints[:]
+
+    substitutions = []
+
+    for same_chip_constraint in constraints:
+        if not isinstance(same_chip_constraint, SameChipConstraint):
+            continue
+
+        # Skip constraints which don't actually merge anything...
+        if len(same_chip_constraint.vertices) <= 1:
+            continue
+
+        # The new (merged) vertex to replace the constrained vertices with
+        merged_vertex = MergedVertex(same_chip_constraint.vertices)
+        substitutions.append(merged_vertex)
+
+        # Remove the merged vertices from the set of vertices resources and
+        # accumulate the total resources consumed. Note add_resources is not
+        # used since we don't know if the resources consumed by each vertex are
+        # overlapping.
+        total_resources = {}
+        for vertex in same_chip_constraint.vertices:
+            resources = vertices_resources.pop(vertex)
+            for resource, value in iteritems(resources):
+                total_resources[resource] = (total_resources.get(resource, 0) +
+                                             value)
+        vertices_resources[merged_vertex] = total_resources
+
+        # A set containing the set of vertices to be merged (for efficient
+        # lookup purposes)
+        merged_vertices = set(same_chip_constraint.vertices)
+
+        # Update any nets which pointed to a merged vertex
+        for net_num, net in enumerate(nets):
+            net_changed = False
+
+            # Change net sources
+            if net.source in merged_vertices:
+                if not net_changed:
```

Comment:
This is correct... I think I probably realised this when I wrote it and kept it on some odd basis but I definitely removed it along with implementing some of your other suggestions... :P